include/autogradpp/detail.h

```diff
@@ -33,11 +35,80 @@ extern tag::Engine engine;
 class ContainerImpl;
 class OptimizerImpl;
 using Variable = tag::Variable;
-using variable_list = tag::variable_list;
 using Tensor = at::Tensor;
 using Container = std::shared_ptr<ContainerImpl>;
 using Optimizer = std::shared_ptr<OptimizerImpl>;
 
+#define GEN_TYPE(TYP, NAME) \
+  Variant(TYP);             \
+  bool is ## NAME () const; \
+  TYP get ## NAME () const;
+
+class Variant {
+ public:
+  Variant(Tensor);
+  Variant(Variable);
+  Variant(const std::string&);
+  Variant(std::vector<Variant>&);
+  Variant(std::vector<Variant>&&);
+  Variant(std::initializer_list<Variable>);
+  Variant(std::unordered_map<std::string, Variant>&);
+  Variant(std::unordered_map<std::string, Variant>&&);
+
+  Variable const&                                       get() const;
+  std::string const&                                    getString() const;
+  std::vector<Variant> const&                           getList() const;
+  std::unordered_map<std::string, Variant> const&       getDict() const;
+  bool                                                  isVariable() const;
+  bool                                                  isString() const;
+  bool                                                  isList() const;
+  bool                                                  isDict() const;
```

Comment:
😍 <- My eyes

Reply:
One question is if doing this is preferred to just .get<Variable>(), which is more clear but may be more typing.

Reply:
For most code it would be `get<ag::Variable>()`, `get<std::string>()`, `get<std::vector<ag::Variable>()`, .... the typing quickly becomes annoying IMHO. I prefer the functions here.

Reply:
(I was referring to the function alignment with my initial comment)